// @langchain/classic/dist/memory/index.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const require_runtime = require("../_virtual/_rolldown/runtime.cjs");
const require_prompt = require("./prompt.cjs");
const require_memory_chat_memory = require("./chat_memory.cjs");
const require_summary = require("./summary.cjs");
const require_summary_buffer = require("./summary_buffer.cjs");
const require_buffer_memory = require("./buffer_memory.cjs");
const require_buffer_window_memory = require("./buffer_window_memory.cjs");
const require_vector_store = require("./vector_store.cjs");
const require_entity_memory = require("./entity_memory.cjs");
const require_combined_memory = require("./combined_memory.cjs");
const require_buffer_token_memory = require("./buffer_token_memory.cjs");
let _langchain_core_messages = require("@langchain/core/messages");
let _langchain_core_chat_history = require("@langchain/core/chat_history");
let _langchain_core_memory = require("@langchain/core/memory");
//#region src/memory/index.ts
var memory_exports = /* @__PURE__ */ require_runtime.__exportAll({
	BaseChatMemory: () => require_memory_chat_memory.BaseChatMemory,
	BaseConversationSummaryMemory: () => require_summary.BaseConversationSummaryMemory,
	BaseMemory: () => _langchain_core_memory.BaseMemory,
	BufferMemory: () => require_buffer_memory.BufferMemory,
	BufferWindowMemory: () => require_buffer_window_memory.BufferWindowMemory,
	ChatMessageHistory: () => _langchain_core_chat_history.InMemoryChatMessageHistory,
	CombinedMemory: () => require_combined_memory.CombinedMemory,
	ConversationSummaryBufferMemory: () => require_summary_buffer.ConversationSummaryBufferMemory,
	ConversationSummaryMemory: () => require_summary.ConversationSummaryMemory,
	ConversationTokenBufferMemory: () => require_buffer_token_memory.ConversationTokenBufferMemory,
	ENTITY_MEMORY_CONVERSATION_TEMPLATE: () => require_prompt.ENTITY_MEMORY_CONVERSATION_TEMPLATE,
	EntityMemory: () => require_entity_memory.EntityMemory,
	VectorStoreRetrieverMemory: () => require_vector_store.VectorStoreRetrieverMemory,
	getBufferString: () => _langchain_core_messages.getBufferString,
	getInputValue: () => _langchain_core_memory.getInputValue,
	getOutputValue: () => _langchain_core_memory.getOutputValue
});
//#endregion
exports.BaseChatMemory = require_memory_chat_memory.BaseChatMemory;
exports.BaseConversationSummaryMemory = require_summary.BaseConversationSummaryMemory;
Object.defineProperty(exports, "BaseMemory", {
	enumerable: true,
	get: function() {
		return _langchain_core_memory.BaseMemory;
	}
});
exports.BufferMemory = require_buffer_memory.BufferMemory;
exports.BufferWindowMemory = require_buffer_window_memory.BufferWindowMemory;
Object.defineProperty(exports, "ChatMessageHistory", {
	enumerable: true,
	get: function() {
		return _langchain_core_chat_history.InMemoryChatMessageHistory;
	}
});
exports.CombinedMemory = require_combined_memory.CombinedMemory;
exports.ConversationSummaryBufferMemory = require_summary_buffer.ConversationSummaryBufferMemory;
exports.ConversationSummaryMemory = require_summary.ConversationSummaryMemory;
exports.ConversationTokenBufferMemory = require_buffer_token_memory.ConversationTokenBufferMemory;
exports.ENTITY_MEMORY_CONVERSATION_TEMPLATE = require_prompt.ENTITY_MEMORY_CONVERSATION_TEMPLATE;
exports.EntityMemory = require_entity_memory.EntityMemory;
exports.VectorStoreRetrieverMemory = require_vector_store.VectorStoreRetrieverMemory;
Object.defineProperty(exports, "getBufferString", {
	enumerable: true,
	get: function() {
		return _langchain_core_messages.getBufferString;
	}
});
Object.defineProperty(exports, "getInputValue", {
	enumerable: true,
	get: function() {
		return _langchain_core_memory.getInputValue;
	}
});
Object.defineProperty(exports, "getOutputValue", {
	enumerable: true,
	get: function() {
		return _langchain_core_memory.getOutputValue;
	}
});
Object.defineProperty(exports, "memory_exports", {
	enumerable: true,
	get: function() {
		return memory_exports;
	}
});
