// @langchain/classic/dist/chains/openai_functions/index.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const require_runtime = require("../../_virtual/_rolldown/runtime.cjs");
const require_extraction = require("./extraction.cjs");
const require_tagging = require("./tagging.cjs");
const require_openapi = require("./openapi.cjs");
const require_base = require("./base.cjs");
//#region src/chains/openai_functions/index.ts
var openai_functions_exports = /* @__PURE__ */ require_runtime.__exportAll({
	createExtractionChain: () => require_extraction.createExtractionChain,
	createExtractionChainFromZod: () => require_extraction.createExtractionChainFromZod,
	createOpenAIFnRunnable: () => require_base.createOpenAIFnRunnable,
	createOpenAPIChain: () => require_openapi.createOpenAPIChain,
	createStructuredOutputRunnable: () => require_base.createStructuredOutputRunnable,
	createTaggingChain: () => require_tagging.createTaggingChain,
	createTaggingChainFromZod: () => require_tagging.createTaggingChainFromZod
});
//#endregion
exports.createExtractionChain = require_extraction.createExtractionChain;
exports.createExtractionChainFromZod = require_extraction.createExtractionChainFromZod;
exports.createOpenAIFnRunnable = require_base.createOpenAIFnRunnable;
exports.createOpenAPIChain = require_openapi.createOpenAPIChain;
exports.createStructuredOutputRunnable = require_base.createStructuredOutputRunnable;
exports.createTaggingChain = require_tagging.createTaggingChain;
exports.createTaggingChainFromZod = require_tagging.createTaggingChainFromZod;
Object.defineProperty(exports, "openai_functions_exports", {
	enumerable: true,
	get: function() {
		return openai_functions_exports;
	}
});
